scripts: get_maintainer.py RFC + provisional MAINTAINERS.yml

ulfalizer opened the pull request:
A script/library I've been working on with input from aescolar (who also
wrote most of the MAINTAINERS.yml file).

Please try it out and give feedback. There's a detailed --help text, and
the top of MAINTAINERS.yml explains the file format.

MAINTAINERS.yml needs to be fleshed out. To be done in future PRs after this one (skeleton) is merged

This probably won't be the final location of the script, but making a PR
against the Zephyr repo makes it easier to try out and update.

Some examples:

* Identify the subsystem(s) for a path:

      $ ./get_maintainer.py path arch/arc/Kconfig
      ARC arch
              status: maintained
              maintainers: ruuddw
              collaborators: vonhust
              inform:
              labels: area: ARC
              description:

      Kconfig
              status: maintained
              maintainers: ulfalizer
              collaborators:
              inform:
              labels: area: Kconfig

* Identify the subsystem(s) for a commit range:

      $ ./get_maintainer.py commits HEAD~5..
      Flash
              status: maintained
              maintainers: nvlsianpu
              collaborators:
              inform:
              labels: area: Flash
              description:

      Kconfig
              status: maintained
              maintainers: ulfalizer
              collaborators:
              inform:
              labels: area: Kconfig

* List all files in a subsystem:

      $ ./get_maintainer.py list "ARC arch"
      arch/arc/CMakeLists.txt
      arch/arc/Kconfig
      arch/arc/core/CMakeLists.txt
      arch/arc/core/arc_connect.c
      arch/arc/core/arc_smp.c
      arch/arc/core/cache.c
      ...

* List all orphaned files (files not in any subsystem):

      $ ./get_maintained.py orphaned
      .checkpatch.conf
      .clang-format
      .codecov.yml
      .editorconfig
      .gitattributes
      .github/ISSUE_TEMPLATE/bug_report.md
      .github/ISSUE_TEMPLATE/enhancement.md
      ...

Older versions of the script had flags instead of sub-commands, like
get_maintainer.pl, but it gets a bit messy with subcommand-specific
flags.

ulfalizer committed:
scripts: get_maintainer.py RFC + MAINTAINERS.yml file

A script/library I've been working on with input from aescolar (who also
wrote most of the MAINTAINERS.yml file).

Please try it out and give feedback. There's a detailed --help text, and
the top of MAINTAINERS.yml explains the file format.

MAINTAINERS.yml needs to be fleshed out. Contributions welcome.

This probably won't be the final location of the script, but making a PR
against the Zephyr repo makes it easier to try out and update.

Some examples:

* Identify the area(s) for a path:

      $ ./get_maintainer.py path arch/arc/Kconfig
      ARC arch
              status: maintained
              maintainers: ruuddw
              collaborators: vonhust
              inform:
              labels: area: ARC
              description:

      Kconfig
              status: maintained
              maintainers: ulfalizer
              collaborators:
              inform:
              labels: area: Kconfig

* Identify the area(s) for a commit range:

      $ ./get_maintainer.py commits HEAD~5..
      Flash
              status: maintained
              maintainers: nvlsianpu
              collaborators:
              inform:
              labels: area: Flash
              description:

      Kconfig
              status: maintained
              maintainers: ulfalizer
              collaborators:
              inform:
              labels: area: Kconfig

* List all files in an area:

      $ ./get_maintainer.py list "ARC arch"
      arch/arc/CMakeLists.txt
      arch/arc/Kconfig
      arch/arc/core/CMakeLists.txt
      arch/arc/core/arc_connect.c
      arch/arc/core/arc_smp.c
      arch/arc/core/cache.c
      ...

* List all orphaned files (files not in any area):

      $ ./get_maintained.py orphaned
      .checkpatch.conf
      .clang-format
      .codecov.yml
      .editorconfig
      .gitattributes
      .github/ISSUE_TEMPLATE/bug_report.md
      .github/ISSUE_TEMPLATE/enhancement.md
      ...

Older versions of the script had flags instead of sub-commands, like
get_maintainer.pl, but it gets a bit messy with subcommand-specific
flags.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

aescolar reviewed (approved):
To me this seems like a perfectly fine starting point.
As soon as we have this in we can start working in parallel on the MAINTAINERS.yml file.
(and while we do that we may have some extra feedback on this script)

nashif commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,384 @@
+    labels:
+        - "area: Sanitycheck"
+
+"Settings subsystem":
```
here using "subsystem", below and in other places, we do not, so we should be consistent, just call it "Settings", no need for subsystem.

ulfalizer replied:
No mention of subsystems anywhere now.

nashif commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,384 @@
+# File format
+# ###########
+#
+# "Subsystem title" (the quotes are needed if the title has more than one word):
```
we probably should use "area" instead of subsystem to avoid confusion with what have under "/subsys/*" and other areas"

ulfalizer replied:
Yeah, area is a better name for documentation and stuff too. Changed to use that in the maintainers file and code.

nashif commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,384 @@
+#     description: >-
+#         Plain-English description. Describe what the system is about, from an
+#         outsider's perspective.
+#
```
add a new field for category, for example:
- driver
- board
- soc 
- architecture
- middleware
- ..

This way we can make better judgment who the assignee (the person driving a Pr to a mergeable state)

ulfalizer replied:
Don't know how it'd get used, so need some feedback:

 - Should it be mandatory or optional?

 - Should it be a freeform string, or is there a list of acceptable values?

aescolar replied:
Would it make sense to add this as a further PR after? As it seems it is not too clear yet

nashif replied:
> Would it make sense to add this as a further PR after? As it seems it is not too clear yet

why? once this is merged, it means the syntax is approved, we should not merge this until we have agreed on the structure and syntax. There is no use for this being merged now anyways.

> As it seems it is not too clear yet

Here is the usecase: you submit a driver that is associated with a board, you have 2 owners/maintainers out of that, you can then make a decision to assign to the board owner instead of assigning to the driver, just an example.
Having this hierarchy and knowledge about what are the code changes are (changing a driver vs changing an architecture) can be easily scriptable.

aescolar replied:
> once this is merged, it means the syntax is approved, we

We can always add new fields after.

Anyhow, @nashif , if you would want it now, I guess Ulf would need a bit more of detail to know what to implement: https://github.com/zephyrproject-rtos/zephyr/pull/20591#discussion_r345375329

ulfalizer replied:
> We can always add new fields after

Yeah, I'd prefer to keep things minimal to start with. Never too late to add stuff.

> Anyhow, @nashif , if you would want it now, I guess Ulf would need a bit more of detail to know what to implement: [#20591 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/20591#discussion_r345375329)

Yeah, will need more information on `category` at least (see above).

nashif reviewed (changes requested):
The starting point of the file should be close to final, i.e. have 80% of areas completed based on previous lists we maintain and based on CODEOWNERS file, if we are to do that one by one, it is going to take ages. I suggest to drive the file close to completion offline and then finalize in GH.

aescolar commented:
@nashif About fleshing out the MAINTAINERS file in this PR. I was actually thinking it would be faster to have the main area maintainers fill-in their areas' details. I'd expect quite a few of them be done quite fast after the release. And then step in to fill what is missing.
The reasoning is that it should be much faster for a maintainer to fill in who are the collaborators, and who are just people to be informed (while for us it will be difficult to guess from CODEOWNERS). Also adding file regexes should be trivial for maintainers, while we would probably be forgetting some files and so forth (so maintainers will anyhow need to go fix after).
It may also be that some area entries will need to be escalated to the TSC due to basic disagreements (about somebody claiming to be a maintainer but not being in the spreadsheet, etc)

ulfalizer commented:
> @nashif About fleshing out the MAINTAINERS file in this PR. I was actually thinking it would be faster to have the main area maintainers fill-in their areas' details. I'd expect quite a few of them be done quite fast after the release. And then step in to fill what is missing.
> The reasoning is that it should be much faster for a maintainer to fill in who are the collaborators, and who are just people to be informed (while for us it will be difficult to guess from CODEOWNERS). Also adding file regexes should be trivial for maintainers, while we would probably be forgetting some files and so forth (so maintainers will anyhow need to go fix after).
> It may also be that some area entries will need to be escalated to the TSC due to basic disagreements (about somebody claiming to be a maintainer but not being in the spreadsheet, etc)

Agreed. I'd probably mess it up if I tried to fill it all in by myself, and I'd rather not have it dumped on me.

Most pragmatic solution would be to merge it I think, and then have the people who work in the different areas fill it in. Could always add a comment that explains that it's not used by CI and the like yet.

ulfalizer commented:
Updated `get_maintainers.py` to work when run in subdirectories too (with paths passed on the command-line interpreted as relative to the working dir). `MAINTAINERS.py` is now looked up in the repository toplevel by default, instead of in the current directory.

aescolar committed:
MAINTAINERS: Add many more entries

Added entries to cover everything in:
https://docs.google.com/spreadsheets/d/1QxfZuM1qQIXy_EJldQeS6nErt7DjHw3SZ2vfiKVdo5M/edit#gid=0

Quite many collaborators have been added based on CODEOWNERS
pending review from the areas' maintainers

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

ulfalizer committed:
MAINTAINERS.yml: Make 'status' optional

Make the status field optional, so sub-areas can add more fields
but let the top area define the maintainance status is needed.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
MAINTAINERS.yml: nits

Yada yada.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

aescolar commented:
In the Lion TSC F2F is was discussed having a MAINTAINERS file in the zephyr root which would both allow contributors to know what files as part of what areas, who maintains them, and what is their maintenance status; And also that we should be able to automate some of our GitHub flow against it.
The idea was further refined and agreed in the process improvement forum afterwards.

The format of the MAINTAINERS file and how the process works around it is described at the beginning of the file itself.
And the file is parsed by this get_maintainers.py file. 

@ruuddw @ioannisg @jhedberg @cvinayak @SebastianBoe
You are "TSC approved maintainers" of an area/subsystem as per [this spreadsheet](https://docs.google.com/spreadsheets/d/1QxfZuM1qQIXy_EJldQeS6nErt7DjHw3SZ2vfiKVdo5M/edit#gid=0)
I have pre-filled this MAINTAINERS.yml file with what is in that excel, the CODEOWNERS file, and my understanding of what files belong to what.
Would you be so kind as to check that the information in the entries for your areas/subsystems is correct, and propose any necessary fixes?
Thanks in advance :)

SebastianBoe commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1009 @@
+    # Provisional entry, to be corrected
+    status: maintained
+    maintainers:
+        - SebastianBoe
```
nashif is a co-maintainer and should be listed if I understand this correctly.

SebastianBoe replied:
Otherwise looks good.

aescolar replied:
"collaborators", roger will do

ioannisg commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1009 @@
+    # Provisional entry, to be corrected
+    status: maintained
+    maintainers:
+        - pizi-nordic
```
Needs to be removed

aescolar replied:
Ok, I understood for a slack discussion that pizi-nordic is not anymore working on Zephyr, so therefore he cannot be a maintainer. This would be just out of date information from the spreadsheet.
I will correct it.

ioannisg replied:
Correct, we need to update the spreadsheet as well.

aescolar replied:
I will mark it here explicitly, so the TSC will be able to review and approve the maintainer change.

ioannisg commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1009 @@
+    collaborators:
+        - MaureenHelm
+        - galak
+        - ioannisg
```
not sure if it makes sense for a person to be both collaborator and maintainer

aescolar replied:
Nope, that would be a mistake

ioannisg replied:
Good. There exist some additional candidates, here. I suppose they can be added any time later, with approval from Maureen and Kumar.

aescolar replied:
Yep, I would think it would be better to do that after this is merged after you say, have the maintainers update their entries after discussing with collaborators. Doing it now could make this PR quite long :)

aescolar commented:
Hi @alexanderwachter @anangl @andrewboie @andyross @dbkinder @dcpleung @finikorg @galak @jakubuC @jukkar @MaureenHelm @mbolivar @mikescott @mnkp @nashif @nategraffsifive @nordickrch @nvlsianpu @pfalcon @rlubos @rveerama1 @tbursztyka @ulfalizer @Vudentz 

In the Lion TSC F2F it was discussed having a MAINTAINERS file in the zephyr root which would both allow contributors to know what files as part of what areas, who maintains them, and what is their maintenance status; And also that we should be able to automate some of our GitHub flow against it.
The idea was further refined and agreed in the process improvement forum afterwards.

The format of the MAINTAINERS file and how the process works around it is described at the beginning of the file itself.
And the file is parsed by the `get_maintainers.py` script.

You are "TSC approved maintainers" of an area/subsystem as per this spreadsheet
I have pre-filled this MAINTAINERS.yml file with what is in that excel, the CODEOWNERS file, and my understanding of what files belong to what.
Would you be so kind as to check that the information in the entries for your areas/subsystems is correct, and propose any necessary fixes?
Thanks in advance :)

alexanderwachter commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1008 @@
+    status: maintained
+    maintainers:
+        - alexanderwachter
+    files:
```
collaborators:
    - karstenkoenig
    - henrikbrixandersen

aescolar replied:
Thanks @alexanderwatcher :). Just a procedural check: @karstenkoenig @henrikbrixandersen : Are you ok with being added here?

henrikbrixandersen replied:
Absolutely. Thank you.

karstenkoenig replied:
Totally, thanks!

alexanderwachter commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1008 @@
+
+"Drivers: HW Info":
+    # Provisional entry, to be corrected
+    status: orphaned
```
status: maintained

alexanderwachter replied:
--- collaborators
+++ maintainer

aescolar replied:
Thanks :)
(It is indeed very nice to have people wanting to be listed as official maintainers)
Note that following process a change like this requires TSC approval, as it is a change from the spreadsheet. (I'm certainly fine with having you here, and I do not think anybody will object in any way to it). So if you do not mind let's take it as a second step, so the TSC will be able to vote on it.

alexanderwachter replied:
Do you suggest to have a vote on it soon or change it in another PR?

aescolar replied:
I was thinking of collecting all these inputs that require a TSC vote together during the following days, and then bring it to a TSC vote (or a set of very fast ones).
Depending on the progress of this PR, it may be that we just do the change here directly, or if the PR was already merged, as a following PR.

henrikbrixandersen commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1008 @@
+        - samples/drivers/entropy/
+        - tests/drivers/entropy/
+
+"Drivers: Entropy: NXP RV32M1":
```
This should probably be `OpenISA RV32M1` instead of `NXP RV32M1`.

aescolar replied:
Thanks :)

henrikbrixandersen commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1008 @@
+        - include/dt-bindings/dma/
+    labels:
+        - "area: DMA"
+
```
The list lacks an entry for `Drivers: EEPROM`.

aescolar replied:
Thanks (This started just based on [this spreadsheet](https://docs.google.com/spreadsheets/d/1QxfZuM1qQIXy_EJldQeS6nErt7DjHw3SZ2vfiKVdo5M/edit#gid=0) which is certainly not so up to date. And that is one of the reasons to have this in the tree).
Should that entry just cover `drivers/eeprom/` and `include/drivers/eeprom.h` ? maybe you want to paste here a diff to apply to the file?

henrikbrixandersen replied:
```"
+ Drivers: EEPROM":
+    # Provisional entry, to be corrected
+    status: orphaned
+    collaborators:
+        - henrikbrixandersen
+    files:
+        - drivers/eeprom/
+        - dts/bindings/mtd/*eeprom*
+        - include/drivers/eeprom.h
+        - tests/drivers/eeprom/
+    labels:
+        - "area: EEPROM"
```

aescolar replied:
Would you offer yourself to be a maintainer for it?

henrikbrixandersen replied:
Yes.

alexanderwachter commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1008 @@
+    labels:
+        - "area: Sensors"
+
+"Drivers: SPI Timer":
```
--- Timer

erwango commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1025 @@
+        - tests/subsys/shell/
+    labels:
+        - "area: Shell"
+
```
"STM32":
    status: maintained
    maintainers:
        - erwango
    files:
        - boards/arm/nucleo_*
        - boards/arm/stm32*_disco
        - boards/arm/stm32*_eval
        - drivers/*/*stm32*
        - drivers/*/*/*stm32*
        - dts/arm/st
        - dts/bindings/*/*stm32*
        - soc/arm/st_stm32
    labels:
        - "platform: STM32"
    description: >-
        STM32 SOCs, dts files and related drivers. ST nucleo, disco and eval boards.

alexanderwachter replied:
This addition would cause lots of drivers to have more than one maintainer.

aescolar replied:
Thank you very much @erwango. Having some components being covered by more than 1 maintainer in different dimensions is, in principle, not a problem (The maintainers would mostly just need to be civil with each other. If they can agree about the direction those common files should go everything should be fine. If they can't, matters can be escalated to the TSC, though hopefully that should not be the case too often).
But I would not encourage having too many maintainers of the same thing to avoid ending too often in deadlocks and needing to escalate to the TSC.
In any case, following process, this entry in itself would need also TSC approval (as unfortunately there is nothing equivalent in the spreadsheet), so I add it to the list.
Loud thinking: It seems I may need to keep some of this "for TSC approval" patches as separate commits.

erwango replied:
> Having some components being covered by more than 1 maintainer in different dimensions is, in principle, not a problem.

This is indeed the current way of working for current STM32 drivers, where I am ensuring that reviews of all STM32 drivers are done either by the driver codeowner (and in that case I may provide a co-review but let the final approval to the codeowner) either by known driver users and myself. This has been the case for several years now and I don't think there are much complaints on this way of working.

alexanderwachter replied:
>and in that case I may provide a co-review but let the final approval to the codeowner

For being informed and providing feedback, a collaborator would be enough right?

Generally, I have no objections adding @erwango as the STM32  maintainer. I just want to mention that conflicts may arise.

Do we add maintainers or collaborators for the other SoCs too?

aescolar replied:
> Do we add maintainers or collaborators for the other SoCs too?

I think we should. But we don't have entries yet.

erwango replied:
> For being informed and providing feedback, a collaborator would be enough right?

This is not the only point, it also consists in ensuring consistency in whole STM32 drivers, dependencies with STM32Cube (when applying) or other drivers (like clock_control today, and potentially pinmux tomorrow, power management, ...).
Last, it may also consists in identifying/promoting/acknowledging maintainers on new drivers or identifying drivers that are no longer maintained (unfortunate, but in the long range, this kind of things happen) and trigger consecutive actions.

alexanderwachter replied:
@erwango I totally agree with you and appreciate having someone doing that work. My concerned are regarding the script that will assign the assignee in the future. The assignee should be the subsys maintainer. Maybe a platform maintainer role could help solving this.
@aescolar what do you think? Should this be discussed in the next process meeting? It would really be nice if we have maintainers for all platforms.

erwango commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1025 @@
+        - tests/subsys/shell/
+    labels:
+        - "area: Shell"
+
```
```suggestion
Shields:
    status: maintained
    maintainers:
        - erwango
    collaborators:
        #To be confirmed
    files:
        - boards/shields
    labels:
        - "area: Shields"
```

@avisconti, @jfischer-phytec-iot would you be interested to be added as collaborators for `boards/shields`?

avisconti replied:
> @avisconti, @jfischer-phytec-iot would you be interested to be added as collaborators for `boards/shields`?

yes sure!

jfischer-phytec-iot replied:
yes

aescolar replied:
Thanks guys, updated now.

erwango replied:
@mike-scott, I forgot you on this. Would you be interested to be added as collaborator for `board/shields`?

erwango replied:
Unresolve, waiting for @mike-scott feedback

aescolar committed:
(TSC) MAINTAINERS: Add "drivers: HW Info" maintainer

alexanderwachter as proposed himself as maintainer for
this area.

TSC approval required

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
(TSC) MAINTAINERS: Add entry for STM32

Adding entry for the STM32 SOCs, boards and drivers

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
(TSC) MAITAINERS: Add entry for Shields

Adding an entry for shields

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar commented:
I'm separating the entries which need their own TSC approval (new areas and/or new maintainers compared with the spreadsheet) in their own commits so it should hopefully be easier to review by the TSC.

jfischer-phytec-iot commented on MAINTAINERS.yml:
```diff
@@ -193,9 +194,65 @@ Bluetooth:
 "Display drivers":
     # Provisional entry, to be corrected
     status: orphaned
+    collaborators:
```
status: maintained
maintainers:
 -vanwinkeljan
collaborators:
 -jfischer-phytec-iot

FIY: @vanwinkeljan

aescolar replied:
Yep, we need @vanwinkeljan confirmation that he wants to be the maintainer :)

vanwinkeljan replied:
> Yep, we need @vanwinkeljan confirmation that he wants to be the maintainer :)

Confirmed

Also add:
* include/display/
* include/drivers/display.h

@jfischer-phytec-iot wondering what to with `lib/gui/` and `subsys/fb/`?
I would add them under display for now.

jfischer-phytec-iot replied:
okay

dcpleung commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1052 @@
+    maintainers:
+        - dcpleung
+    files:
+        - drivers/serial/
```
Also `dts/bindings/serial/`

aescolar replied:
Thanks @dcpleung. Note that there was an issue/inconsistency regarding this entry: the CODEOWNERS file does not seem to contain any line assigned to you for these file. 
But I take your feedback as you agreeing otherwise about this entry.

dcpleung replied:
I am happy to look after this component. :)

dcpleung commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1052 @@
+        - andrewboie
+    files:
+        - arch/xtensa/
+        - include/arch/xtensa/
```
Also `dts/xtensa/`

mbolivar commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1052 @@
+    description: >-
+        Inter-processor mailboxes
+
+"Drivers: I2C":
```
nope, I don't maintain this

aescolar replied:
Good that you spotted it. Thanks.

mbolivar commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1052 @@
+        - "area: LED"
+
+"Drivers: LED Strip":
+    # Provisional entry, to be corrected
```
confirmed; I am the maintainer

mbolivar commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1052 @@
+        - "area: Userspace"
+
+West:
+    # Provisional entry, to be corrected
```
confirmed; I am the maintainer. @carlescufi as collaborator please

aescolar replied:
@carlescufi : Procedural check: Can you confirm that you want to be listed as collaborator for west?

carlescufi replied:
Yes please

mbolivar reviewed (commented):
@aescolar I appreciate your efforts on this. I have taken a look at areas where I was mentioned and confirmed or denied.

rlubos commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1052 @@
+        - samples/net/lwm2m_client/
+        - subsys/net/lib/coap/
+        - subsys/net/lib/lwm2m/
+        - subsys/net/lib/openthread/
```
Shall we add a separate entry for OpenThread then?

aescolar replied:
CC @jukkar.
This I wrote due to the current rules in CODEOWNERS. But I certainly need help from you guys here.
Regarding your exact question: It is possible (Maybe you want to take a look at the BT entries as an example of another complex area). There is the option of splitting some part fully into its own area. Or also of adding a sub-area to an area, where that sub-area may have its own maintainer, or just extend with some extra collaborators.

nvlsianpu commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1052 @@
+    files:
+        - include/settings/
+        - subsys/settings/
+        - tests/subsys/settings/
```
```suggestion
        - tests/subsys/settings/
        - doc/reference/runtime_conf/
```

erwango commented:
@aescolar, one point is not clear to me. Does this fully replace `CODEOWNERS` file and if not, how will they co-exist?

aescolar commented:
> @aescolar, one point is not clear to me. Does this fully replace CODEOWNERS file and if not, how will they co-exist?

@erwango The hope is indeed to replace CODEOWNERS with this, but it takes several steps to get there:
* This PR needs to be accepted and the TSC needs to approve the changes/additions of MAINTAINERS we have on it.
* With the file in, we can add the automation. It would need to coexist with CODEOWNERS for a small while.
* We need to fill in all the collaborators and `inform` entries (which is already happening, and will probably happen with a few subsequent PRs)
* Whenever we have everything in, and everything seems to be working stably, we can do the switch.

aescolar commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1025 @@
+
+DFU:
+    # Provisional entry, to be corrected
+    # TODO: spreadsheet not matching CODEOWNERS
```
The comment was because I did not see entries in the CODEOWNERS file covering these.
So, you agree with entry?

avisconti commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1052 @@
+    maintainers:
+        - MaureenHelm
+    files:
+        - drivers/sensor/
```
I maintain all the ST sensors:

/drivers/sensor/hts*/  
/drivers/sensor/lis*/ 
/drivers/sensor/lps*/  
/drivers/sensor/lsm*/  
/drivers/sensor/st*/ 

Not sure how to handle this.
Maybe you can put me as collaborator if Maureen agrees.

aescolar replied:
Thanks @avisconti . The best thing is of course to agree with @MaureenHelm. Maybe you want to have a new area entry with those(?). Then you could either be set as maintainer of that sub-area, or just a collaborator. You just tell me the changes you want, and I add them here as a separate commit.

avisconti replied:
@aescolar 
I think it should be difficult to separate the ST part from the rest (not sure '*' is allowed).
Probably Collaborator of the 'sensor' part (driver, samples, ...) would be fine.
Let's see what @MaureenHelm has to say.

aescolar replied:
> (not sure '*' is allowed)

It certainly is :). You may want to look into one of the entries or more complex subsystems today, like BT, for some examples of what is possible

avisconti replied:
@aescolar 
Verified with @MaureenHelm .
You can put me as Collaborator for sensors

pfalcon commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1052 @@
+# Changes to MAINTAINERS.yml need to be approved as follows:
+#
+#     * Changing the 'maintainers' for an area needs approval from the
+#       technical steering committee
```
Capitalize words here, so we all know it's ol' good TSC?

aescolar replied:
corrected

pfalcon commented:
>>     @aescolar, one point is not clear to me. Does this fully replace CODEOWNERS file and if not, how will they co-exist?

> @erwango The hope is indeed to replace CODEOWNERS with this, but it takes several steps to get there:

I appreciate the work done on this and very active feedback, but I agree with @erwango that the baseline situation isn't very clear (even to someone who was at the TSC F2F 2019-11, where this matter was discussed). So, perhaps it would be helpful to add to the description summary of what's wrong with CODEOWNERS file and why we need MAINTAINERS file (or short summary and a link to detailed description/discussion). The biggest question for me is whether this comes (will soon come) with specific document which describes responsibilities, rights of maintainers and description of various processes (e.g. conflict resolution, escalation, expected process timeframes, etc.). Thanks.

erwango commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1052 @@
+#         outsider's perspective.
+#
+#
+# All areas must have a 'status' key and a 'files' and/or 'files-regex' key.
```
Status is optional, from what I understood.
Though, it would be needed to illustrate in which conditions it could be ommitted

aescolar replied:
Yep, that was a left over, corrected now

aescolar committed:
(TSC) MAINTAINERS: Add vanwinkeljan as Display maintainer

Add vanwinkeljan as maintainer and jfischer-phytec-iot
as collaborator.
See:
https://github.com/zephyrproject-rtos/zephyr/pull/20591#discussion_r354418992

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
MAINTAINERS: Expand "Display drivers" scope

Expand "Display drivers" scope to include lib/gui/ and subsys/fb/

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
MAINTAINERS: mbolivar is not the I2C maintainer

There seems to have been an error in the old spreadsheet.
mbolivar is not the maintainer of this area
=> changed to orphan

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
(TSC) MAINTAINERS: Add mbolivar as maintainer for LED Strip

Promote mbolivar as maintainer to this area

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

erwango commented on MAINTAINERS.yml:
```diff
@@ -942,6 +942,18 @@ Shell:
+        - jfischer-phytec-iot
+    files:
+        - boards/shields/
+    labels:
```
Forgot this:
```
        - doc/guides/porting/shields.rst
```

erwango commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1055 @@
+        - include/dt-bindings/dma/
+    labels:
+        - "area: DMA"
+
```
```suggestion
"Drivers: DMA stm32":
    # Provisional entry, to be corrected
    collaborators:
        - cybertale
    files:
        - drivers/adc/dma_stm32.c
        - dts/bindings/dma/st,stm32-dma.yaml
        - include/dt-bindings/dma/stm32_dma.h
```
@cybertale , can you confirm ?

erwango commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1050 @@
+        - drivers/ipm/
+    description: >-
+        Inter-processor mailboxes
+
```
```suggestion
"Drivers: IPM stm32":
    # Provisional entry, to be corrected
    collaborators:
        - arnopo
    files:
        - drivers/ipm/ipm_stm32_ipcc.c
        - dts/bindings/ipm/st,stm32-ipcc-mailbox.yaml
```
@arnopo, can you confirm ?
(Btw, you need to update CODEOWNERS with your new GH handle)

arnopo replied:
Yes Ok for me 
Yes need to update CODEOWNERS : https://github.com/zephyrproject-rtos/zephyr/pull/21228

erwango commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1050 @@
+        - include/drivers/i2c/
+    labels:
+        - "area: I2C"
+
```
```suggestion
"Drivers: I2C stm32":
    # Provisional entry, to be corrected
    collaborators:
        - ydamigos
    files:
        - drivers/i2c/i2c_ll_stm32*
```
@ydamigos, can you confirm ?

ydamigos replied:
Yes, Ok for me

erwango commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1050 @@
+        - tests/drivers/i2s/
+    labels:
+        - "area: I2S"
+
```
```suggestion
"Drivers: I2S stm32":
    # Provisional entry, to be corrected
    collaborators:
        - avisconti
    files:
        - drivers/i2s/i2s_ll_stm32*
```
@avisconti, can you confirm ?

avisconti replied:
yes!

erwango commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1050 @@
+        - drivers/timer/
+    labels:
+        - "area: Timer"
+
```
```suggestion
"Drivers: USB STM32":
    # Provisional entry, to be corrected
    collaborators:
        - ydamigos
        - loicpoulain
    files:
        - drivers/usb/device/usb_dc_stm32.c
        - dts/bindings/usb/st,stm32-*
```
@ydamigos, @loicpoulain, can you confirm ?

ydamigos replied:
Yes, Ok for me.

jfischer-phytec-iot replied:
Disagree. Suggest:

```diff
+"Drivers: USB":
+    # Provisional entry, to be corrected
+    status: maintained
+    maintainers:
+        - jfischer-phytec-iot
+    collaborators:
+        - finikorg
+    files:
+        - drivers/usb/device/*
+        - dts/bindings/usb/*
```

erwango replied:
@jfischer-phytec-iot, can you elaborate?
Is there an incompatibility in having maintainer on a driver (as a subsystem) and having collaborators on a particular driver (of this subsystem)?

erwango commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1050 @@
+        - drivers/timer/
+    labels:
+        - "area: Timer"
+
```
```suggestion
"Drivers: LPTIMER STM32":
    # Provisional entry, to be corrected
    collaborators:
        - FRASTM
    files:
        - drivers/timer/stm32_lptim_timer.c
        - dts/bindings/timer/st,stm32-timers.yaml
```
@FRASTM, can you confirm ?

erwango reviewed (changes requested):
Adding entries for stm32 drivers collaborators.

avisconti commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1050 @@
+    description: >-
+        STM32 SOCs, dts files and related drivers. ST nucleo, disco and eval
+        boards.
+
```
BOARDS(???):
    status: maintained
    maintainers:
        - avisconti
    files:
        - boards/arm/96b_argonkey/ 
        - boards/arm/sensortile_box/ 
    labels:
        - 
    description: >-
        ....

erwango replied:
@avisconti, I guess this has to be put in a boards dedicated section

avisconti replied:
@erwango Yes, I guess it is still missing.

ulfalizer committed:
get_maintainer.py: Simplify Area._contains() a bit

Treat a missing files/files-exclude like an empty list.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
get_maintainer.py: Update outdated comment in commits2areas

The -f flag was removed. There's now 'path' and 'commits' subcommands
instead.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
MAINTAINERS.yml: Reflow paragraph in intro

Use 'gq' in Vim.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
MAINTAINERS.yml: Move ZTest entry last

Make alphabetically sorted.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
CODEOWNERS: Add missing nicks

Stops CI check from complaining.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
get_maintainer.py: Speed up 'orphaned' and 'list' ~6x

Construct a single regex for matching files/files-regex, and another
regex for matching files-exclude/files-regex-exclude. This optimizes
Area._contains() down to two regex matches maximum, which are done in C.

Brings 'orphaned' and 'list' down to ~0.4s from ~2.5s on my system, with
a warm file cache.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer commented:
`orphaned`, `list`, and anything else that needs to match a ton of paths are now 6x faster with help from some regex hackery.

jhedberg commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1051 @@
+    collaborators:
+        - trond-snekvik
+    files:
+        - subsys/bluetooth/mesh/
```
Please add `include/bluetooth/mesh/` here

ulfalizer replied:
Done

ioannisg commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1051 @@
+    labels:
+        - "area: ARM"
+
+Bluetooth:
```
Doesn't this need " "?

ulfalizer replied:
Nope, and not needed even if there's a space in it apparently. Pushed a clean-up commit that removes some more quotes.

ioannisg commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1051 @@
+    # TODO: The entries in the maintainers spreadsheet do not match CODEWONERS at all!
+    status: maintained
+    maintainers:
+        - ruuddw
```
are we sure about this (just checking)

mbolivar commented on get_maintainer.py:
```diff
@@ -0,0 +1,532 @@
+See the help texts for the various subcommands for more information. They can
+be viewed with e.g.
+
+    ./get_maintainer.py path --help
```
Hmm, actually, shouldn't we make this a west extension? That way it could be invoked as:

`west get-maintainer path`

or so. Here is the [extension documentation](https://docs.zephyrproject.org/latest/guides/west/extensions.html).

No worries if you want an independently runnable script which is also a library also, that makes sense, but I think since this is user facing, the primary user interface should be west.

ulfalizer replied:
You're free to add a west extension. I think we should move away from the big monolithic tool approach and have simple and easy to debug standalone script instead, so I'd rather not.

mbolivar replied:
> simple and easy to debug standalone scripts instead

Why are they mutually exclusive in this case?

ulfalizer committed:
get_maintainer.py: Remove redundant quotes around titles

Reading up some more, quotes around titles are only needed for titles
with colons and the like in them.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
MAINTAINERS.yml: Add include/bluetooth/mesh/ to Bluetooth Mesh

Requested by jhedberg.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ioannisg commented on MAINTAINERS.yml:
```diff
@@ -0,0 +1,1052 @@
+#
+# Ideally, any file in the tree will be covered by some area.
+#
+# When a GitHub pull request is sent, this happens:
```
the following happens

nashif committed:
MAINTAINERS: remove non-existing files

Remove files not part of the tree anymore.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
(TSC) MAINTAINERS: update entries for doc/ and build system

Updated
- Build System
- Documentation

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

ioannisg committed:
MAINTAINERS: Add stephanosio and carlocaione as collaborators for ARM

Adding carlocaione and stephanosio as collaborators
for the ARM architecture code. Add also the directory
of the ARM-specific tests into the ARM architecture,
as these tests are meant to be ARM specific.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
(TSC) MAINTAINERS: add entry for MAINTAINERS file

Add TSC Chair as the maintainer of the MAINTAINERS.yml file.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
(TSC) MAINTAINERS: add entry for arch/common directory

Add Andrew Boie as the maintainer of the arch/common directory.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
TSC: MAINTAINERS: add more entries as orphaned

Add a list of missing entries of Drivers
and Sub-systems as orphaned.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

nashif commented:
superseded with #24152
